src/main/scala/rules/Consumer.scala

```diff
@@ -199,6 +236,44 @@ object consumer extends ConsumptionRules {
               Q(s2, h, Unit, v2)
             }))
 
+      case ite @ ast.CondExp(e0, a1, a2) if !a.isPure && Verifier.config.moreJoins() =>
+        val condExpRecord = new CondExpRecord(ite, s, v.decider.pcs, "consume")
+        val uidCondExp = SymbExLogger.currentLog().openScope(condExpRecord)
+
+        eval(s, e0, pve, v)((s1, t0, v1) =>
+          joiner.join[(Heap, Term), (Heap, Term)](s1, v1, resetState = false)((s1, v1, QB) => {
+            branch(s1, t0, v1)(
+              (s2, v2) => consumeR(s2, h, a1, pve, v2)((s3, h1, t1, v3) => {
+                SymbExLogger.currentLog().closeScope(uidCondExp)
+                QB(s3, (h1, t1), v3)
+              }),
+              (s2, v2) => consumeR(s2, h, a2, pve, v2)((s3, h1, t1, v3) => {
+                SymbExLogger.currentLog().closeScope(uidCondExp)
+                QB(s3, (h1, t1), v3)
+              }))
+          })(entries => {
+            val s2 = entries match {
+              case Seq(entry) => // One branch is dead
+                (entry.s, entry.data)
+              case Seq(entry1, entry2) => // Both branches are alive
+                val mergedData = (
```

Comment:
Why is `State.mergeHeap` necessary here, whereas just `entry1.pathConditionAwareMerge` suffices in other places, e.g. for performing a CFG join?

Reply:
`pathConditionAwareMerge` can be used to easily merge the State, it just calls the method `merge` from the State and supplies the corresponding branch conditions. But here, we merge the additional data (I think you called it join data, generically annotated `JD` in the joiner) collected in the branches, for example the partially consumed heap, which is needed again in the continuation.

Reply:
Thanks for the technical explanation. What I don't understand, though, is why the additional data is needed here, but not always?

Reply:
Maybe I misunderstood your question, but the consumer requires this additional data for the continuation, the producer does not.

The continuation used in the consumer has the following signature:

```
(Q: (State, Heap, Term, Verifier) => VerificationResult)
```

In the producer, the signature of the continuation is:

```
(Q: (State, Verifier) => VerificationResult)
```

The two arguments of type `Heap` and `Term` are the ones that additionally merged in the consumer, in the producer for example we don't need to do this as this additional data simply does not exist in its implementation.

Reply:
Thanks, now it makes sense. I guess I simply haven't taken the time to look at the context.